### ISSUE TITLE
When including a plugin that also uses custom-config, config.xml configs are overwritten

### ISSUE THREAD

**Issue by ivynye** [dpa99c/cordova-custom-config#51]

**Comment by dpa99c** [comment]
From your description, I don't understand the specific problem. By this, do you mean that another plugin is making changes to a platform config and that these changes are being lost when this plugin applies its changes? Because this plugin doesn't directly write to the config.xml 

If this is the case, please post your config.xml and the verbose console output of the prepare or build operation which illustrates the issue

**Comment by becvert** [comment]
Hi. I think I have the same issue.

For instance, one of my "custom-config" blocks in config.xml is:

```
<config-file parent="LSApplicationQueriesSchemes" platform="ios" target="*-Info.plist">
            <array>
                <string>itms-apps</string>
                <string>twitter</string>
                <string>fb</string>
                <string>belote-rebelote-free</string>
                <string>belote-rebelote-paid</string>
                <string>tarot-africain-free</string>
                <string>tarot-africain-paid</string>
            </array>
        </config-file>
```

And I'm using the Facebook4 cordova plugin which add in the .plist:

```
<key>LSApplicationQueriesSchemes</key>
    <array>
      <string>fbapi</string>
      <string>fb-messenger-api</string>
      <string>fbauth2</string>
      <string>fbshareextension</string>
    </array>
```

so I should have this as the result:

```
<key>LSApplicationQueriesSchemes</key>
    <array>
      <string>itms-apps</string>
      <string>twitter</string>
      <string>fb</string>
      <string>belote-rebelote-free</string>
      <string>belote-rebelote-paid</string>
      <string>tarot-africain-free</string>
      <string>tarot-africain-paid</string>
      <string>fbapi</string>
      <string>fb-messenger-api</string>
      <string>fbauth2</string>
      <string>fbshareextension</string>
    </array>
```

it turns out that I only get this:

```
<key>LSApplicationQueriesSchemes</key>
    <array>
      <string>itms-apps</string>
      <string>twitter</string>
      <string>fb</string>
      <string>belote-rebelote-free</string>
      <string>belote-rebelote-paid</string>
      <string>tarot-africain-free</string>
      <string>tarot-africain-paid</string>
    </array>
```

I must re-install Facebook4 to get what I want.
That plugin is able to merge blocks somehow.
"Custom-config" blocks should merge with existing ones too.

I think it's a possible duplicate of #30  

Thank you. awesome plugin otherwise!

**Comment by dpa99c** [comment]
@becvert from the info you've posted, I can see what's happening: currently this plugin does not parse the contents of config-file blocks, it merely replicates the contents verbatim inside of the specified root node. So if the root node already exists in the platform config (in this case it does as `LSApplicationQueriesSchemes`), the existing values will be replaced by those defined in the config block.

In order to merge rather than replace, the plugin would need to parse and process the config-file block contents as XML nodes rather than plain text.

The Facebook4 plugin has hard-coded values for this root node, so simply [defines them in its plugin.xml](https://github.com/jeduan/cordova-plugin-facebook4/blob/master/plugin.xml#L138) and lets Cordova handle the merge.

If/when I have time, I will make this enhancement to the plugin.

**Comment by becvert** [comment]
great. thank you.

as a workaround for now, I change the hook's type in your plugin.xml from

`<hook src="hooks/applyCustomConfig.js" type="after_prepare" />`

to

`<hook src="hooks/applyCustomConfig.js" type="before_prepare" />`

in order to let the Facebook4 plugin or rather cordova itself do the merge.

**Comment by ivynye** [comment]
It looks like becvert has the same issue. We're using a custom-config block to set some ios preference, here are a few:
```
        <config-file parent="UIStatusBarHidden" platform="ios" target="*-Info.plist">
            <true />
        </config-file>
        <config-file parent="UIViewControllerBasedStatusBarAppearance" platform="ios" target="*-Info.plist">
            <false />
        </config-file>
```
We just recently started to implement the cordova-plugin-tag-manager, which also has some <config-file> tags configured.

I also discovered that timing on the applyCustomConfig hook was the source of the issue.

**Comment by dpa99c** [comment]
@ivynye is your issue exactly the same as @becvert? i.e. the other plugin is creating config entries with the same element names and this plugin is overwriting those? Or is your issue that the other plugin is creating entirely different config entries yet the changes are being lost anyway?

I can understand that in @becvert's case, setting the hook to run `before_prepare` solved the issue because this plugin was overwriting elements that had previous been created by the Facebook plugin.

However, if I set the plugin to always do this by default, then it loses the ability to override defaults that are set by Cordova during the `prepare` operation (running `after_prepare` enables this). One quick solution I could implement is to allow a custom preference in the config.xml which enables this to be specified (e.g. `<preference name="cordova-custom-config-hook" value="before_prepare|after_prepare" />`). Would that solve your issues?

**Comment by Ajaxy** [comment]
+1 here.
Would be greate to have `overwrite` parameter as it's done in PhoneGap Build (tree parse will be needed) — http://phonegap.com/blog/2014/01/30/customizing-your-android-manifest-and-ios-property-list-on-phonegap-build/

**Comment by kwvanderlinde** [comment]
This would be a welcome feature. It seems PhoneGap has replaced the `overwrite` attribute with the `mode` attribute (see docs [here](http://docs.phonegap.com/phonegap-build/configuring/config-file-element/)). I think `mode` is a better parameter since it gives more control than just whether or not to overwrite.

**Comment by pke** [comment]
Hitting the exact same issue as @becvert today with the facebook4 plugin.
Just switching the apply step to `<hook src="hooks/applyCustomConfig.js" type="before_prepare" />` seems fishy to me and to have sideeffects @dpa99c described.

I wish cordova would have provided a annotate config functionality. That way plugins or anybody would get the current config xml and have functions like "merge nodes", "append nodes", what you have in XML manipulation.
Would make all this custom fragile hand parsing obsolete.

One other solution would be to have our own plugin with all the custom settings and install that plugin into the platform and let cordova handle the merge.

**Comment by pke** [comment]
@dpa99c I think the culprit is this line https://github.com/dpa99c/cordova-custom-config/blob/master/hooks/applyCustomConfig.js#L483

There you override whatever is in the plist already.
Couldn't that be:
`infoPlist[key] = Object.assign({},  infoPlist[key], value)`

**Comment by dpa99c** [comment]
@pke 
> I wish cordova would have provided a annotate config functionality. That way plugins or anybody would get the current config xml and have functions like "merge nodes", "append nodes", what you have in XML manipulation.
Would make all this custom fragile hand parsing obsolete.

I wish the same. I created this plugin out of my own necessity, and it's becoming more and more of a pain to maintain. Ironically, I hardly use it myself anymore, so it's for the goodness of the Cordova community that I don't just unpublish the damned thing.

The support for [`<edit-config>` blocks](http://cordova.apache.org/docs/en/6.x/plugin_ref/spec.html#edit-config) in config.xml [added in cordova@6.4.0](https://cordova.apache.org/news/2016/10/25/tools-release.html)  is a step in the right direction, but still only works for AndroidManifest.xml - no .plist support.
And `<config-file>` blocks are still not supported in config.xml. If Cordova supported all the same elements in config.xml as it does in plugin.xml, this plugin could be locked away in the basement.

**Comment by pke** [comment]
@dpa99c PR #92 if you would be so kind :)

**Comment by pke** [comment]
@dpa99c Have fixed this one now for good with #93

**Comment by edenworky** [comment]
Still running into this issue with the Facebook plugin on Android. Has a solution been found using `<edit-config>`? It's not apparent to me.
